---
title: "polyBreedR Vignette 3: Introduction to Genomic Selection"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{polyBreedR Vignette3}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE,message=FALSE,warning=FALSE,comment="##",
                      fig.width=4,fig.height=4,dpi=150)
```

### Introduction

Genomic selection (GS) leverages historical phenotype data from a breeding program to more accurately predict the genetic value of related individuals. The most important application of GS is the prediction of additive value, so that superior parents can be selected more quickly than using phenotypes alone. This is because reducing the length of the breeding cycle (not to be confused with the time to variety release) is one of the most powerful drivers of genetic gain ([Cobb et al. 2019](https://doi.org/10.1007/s00122-019-03317-0)).

Historical data are organized as distinct trials that occur at a particular place and time, which are called environments in plant breeding. Proper biometric procedures are needed to combine all of this information while accounting for micro- and macro-environmental variation. The polyBreedR package uses a fully efficient, two-stage procedure ([Damesa et al. 2017](https://doi.org/10.2134/agronj2016.07.0395)) that requires some entries to be replicated in each environment. In Stage 1, each environment is analyzed separately, taking into account the experimental design and (if desired) any spatial trends. Best linear unbiased estimates (BLUEs) of the genotypic (i.e., clonal) value of the individuals are computed, as well as the variance-covariance (vcov) matrix of the BLUEs. Both pieces of information are used in Stage 2 to partition the variance of the BLUEs into additive and non-additive genetic effects and genotype x environment interactions.

### Phenotype Data

The tutorial dataset is based on six years (2015-2020) of potato breeding trials at the Hancock Research Station of the University of Wisconsin (UW), which is an updated version of the dataset published by [Endelman et al. (2018)](https://doi.org/10.1534/genetics.118.300685). A mixture of augmented and partially replicated designs have been used with incomplete blocks. Four traits are included: 

* total yield ($Mg$ $ha^{-1}$) 

* vine maturity (1=early to 9=late)

* specific gravity

* fry color, measured in units of Hunter Lightness ($L$) after 6 months of storage

The data file also contains the stand count for each plot (out of 15 total plants), which is included as a covariate in the Stage 1 model. 

```{r, fig.width=4,fig.height=3}
pheno.file <- system.file("vignette_data", "GS_pheno.csv", package = "polyBreedR")
pheno <- read.csv(pheno.file)
head(pheno)

library(ggplot2)
ggplot(data=pheno,aes(x=env)) + geom_bar() + ylab("Number of plots") + 
  xlab("Environment") + theme(axis.text.x=element_text(angle=45,hjust=1,size=10))
```

As is typical of breeding trials, the majority of clones were tested in one year and then dropped, but there is sufficient replication across years to estimate genotype x year interactions.
```{r}
tab <- table(pheno$id,pheno$env)
env.per.id <- apply(tab,1,function(x){sum(x>0)})
replicated <- names(env.per.id)[env.per.id > 1]

#Number of clones tested in more than one env
sapply(tapply(pheno$id,pheno$env,intersect,y=replicated),length)
```


### Genotype Data

Our genotyping platform for tetraploid potato has been an Infinium SNP array. Most of the clones in this dataset were genotyped with Version 3 (V3) of the array, but some were genotyped with V2, and some have no marker data for various reasons. (See [Vignette 1](https://jendelman.github.io/polyBreedR/polyBreedR_Vignette1.html) for more information about genotype calling with polyBreedR.) When two groups of individuals have been genotyped on different platforms with many shared markers, the polyBreedR function `merge_impute` can be used to merge the datasets and impute the missing values by BLUP. The algorithm in `merge_impute` is a modification of the EM algorithm in function `A.mat` of the rrBLUP package, but the polyBreedR implementation is not iterative and is not recommended for use with RADseq/GBS data. 

```{r,fig.width=4,fig.height=3}
genoV2.file <- system.file("vignette_data", "GS_genoV2.csv", package = "polyBreedR")
genoV3.file <- system.file("vignette_data", "GS_genoV3.csv", package = "polyBreedR")
genoV2 <- as.matrix(read.csv(genoV2.file,row.names=1,check.names=F))
genoV3 <- as.matrix(read.csv(genoV3.file,row.names=1,check.names=F))
dim(genoV2) #markers x indiv
dim(genoV3) #markers x indiv

#number of markers in common
length(intersect(rownames(genoV2),rownames(genoV3)))

library(polyBreedR)
geno <- merge_impute(genoV2,genoV3,ploidy=4)
dim(geno)
ggplot(data.frame(x=as.vector(geno)),aes(x=x)) + geom_histogram() + xlab("Allele Dosage")
```

As the above figure illustrates, the imputation process generates fractional values of dosage, but these are compatible with estimation of the G matrix for predicting additive values, using polyBreedR function `G_mat`.
```{r,fig.height=3,fig.width=4}
G.ans <- G_mat(geno,ploidy=4)
G <- G.ans$mat  #G matrix

ggplot(data.frame(G=as.vector(G[upper.tri(G,diag=F)])),aes(x=G)) + geom_histogram()
```

The above figure illustrates the typical, unimodal, zero-mean distribution of G matrix coefficients for a homogeneous population.

### Two-stage analysis

polyBreedR functions `Stage1` and `Stage2` implement the two-stage analysis methodology described earlier. Both require a license for ASReml-R (version 4.1.0.148 or later), which can be obtained from [VSN International](https://www.vsni.co.uk/software/asreml-r). The `Stage1` command expects columns named "id" and "env" in the input data frame, as well as columns for each trait. The data for each environment x trait combination are analyzed independently with a linear mixed model. Any additional fixed or random effects in the model are specified using "effects", which is a named list of 2-vectors as shown in the example below. For each effect, the first element must be "fixed" or "random", and the second element is "factor" or "numeric" (i.e., covariate). `Stage1` returns a list with three elements:

* "H2" is a matrix of the plot-based, broad-sense heritability for each env x trait combination

* "blue" is a data frame of BLUEs for each individual x trait combination

* "blue.vcov" is a list of matrices (one for each trait) with the information needed for Stage 2


```{r,fig.height=4}
library(asreml)
ans1 <- Stage1(data=pheno,
               traits=c("total.yield","vine.maturity","spec.grav","fry.color"),
               effects=list(block=c("fixed","factor"),stand.count=c("fixed","numeric")),
               silent=TRUE)
head(ans1$H2)

library(reshape2)
plot.data <- melt(ans1$H2)
ggplot(plot.data,aes(x=Var2,y=value,colour=Var1)) + geom_point() + xlab("") + 
  scale_y_continuous(expression(H^2),limits=c(0,1),breaks=seq(0,1,by=0.2),labels=seq(0,1,by=0.2)) + 
  guides(colour=guide_legend(title="Env")) + theme(axis.text.x=element_text(angle=45,hjust=1,size=10))
```

The above figure shows the distribution of plot-based $H^2$ for the 6 enviroments. 

Clones without marker data need to be removed from the Stage 1 results before continuing to Stage 2, which is one of the features of `Stage2_prep`. This function also returns a matrix called Omega, which must be saved as a variable in the global environment for ASReml-R to work properly in `Stage2`. The same is true for the G matrix (created above) and other covariance matrices, which are called "kernels" in `Stage2`. The `Stage2` function automatically includes a random genetic effect with an identity (I) variance-covariance matrix, which represents non-additive effects when used in combination with a random effect with covariance proportional to G. At present, each trait must be analyzed separately in Stage 2, but a multivariate analysis to estimate the genetic correlation between traits is under development. Here is the code to analyze total yield:
```{r}
not.genotyped <- setdiff(ans1$blue$id,rownames(G))
ans1b <- Stage2_prep(ans1$blue.vcov["total.yield"],exclude.id=not.genotyped)
Omega <- ans1b$Omega
blue <- ans1b$blue
ans2 <- Stage2(data=blue,traits="total.yield",kernels="G",silent=F)
round(ans2$vars,1)
```

The output "vars" contains the estimated variance for each effect, and the results for "kernel=G" and "kernel=I" correspond to additive and non-additive genetic effects, respectively. (For kernels, the variance component is multiplied by the mean of the diagonal elements.) In this example, GxE refers to the genotype x year interaction because the data come from a single location. (A tutorial for multi-location datasets is under development.) 
```{r}
#proportion of variance
ans2$vars/sum(ans2$vars)
```

Genomic narrow-sense heritability, defined here as the proportion of variance for the additive effects, is estimated at 0.31 for total yield (on a plot basis).

The `Stage1` function requires partial replication within each environment to estimate the variance-covariance matrix of the BLUEs. If your dataset does not satisfy this requirement, there is the option of running `Stage2` without the Omega matrix, which is suboptimal because the micro- and macro-environmental variation are confounded. The benefit of including Omega in the analysis is reflected in a lower AIC compared to a model without it.

```{r}
remove("Omega") #remove variable from the global environment
ans2b <- Stage2(data=blue,traits="total.yield",kernels="G",silent=T)

#Model with Omega has lower AIC:
ans2$aic - ans2b$aic
```


### Genomic prediction by BLUP

One additional function, `predict_MME`, is needed to predict the breeding values in polyBreedR. As the name suggests, it is based on solving the Mixed Model Equations (MME). The necessary data for `predict_MME` are encapsulated in an S4 class called `MME`, which is returned by `Stage2` (but can also be created explicitly by advanced users). Inversion of the MME coefficient matrix enables computation of the reliability ($r^2$) of the BLUPs, which is the expected squared correlation between the true and predicted values. This quantity is analogus to narrow-sense $h^2$ and is an alternative to cross-validation (CV) for assessing genomic prediction accuracy. Cross-validation tends to underestimate prediction accuracy because the data used for validation are not true breeding values (to compensate, the correlation can be divided by narrow-sense $h$). On the other hand, reliability from the MME tends to overestimate accuracy because it assumes the model is correct.

The argument "weights" of `predict_MME` allows the user to specify the linear combination of genetic effects to predict. For parent selection based on additive values, use weights of 0 and 1 for the I and G kernels, respectively. To select on clonal values for variety release, both kernels should have weights of 1.
```{r,fig.height=4,fig.width=4}
add.pred <- predict_MME(data=ans2$MME,weights=c(I=0,G=1))
head(add.pred[order(add.pred$blup,decreasing=T),]) #Top yielders

clonal.pred <- predict_MME(data=ans2$MME,weights=c(I=1,G=1))

combo <- merge(clonal.pred,add.pred,by="id")

ggplot(data=combo,aes(x=blup.x,y=blup.y)) + geom_point() + ylab(expression("Additive Yield"~(Mg~ha^-1))) + 
  xlab("Clonal Yield"~(Mg~ha^-1)) + geom_abline(slope=1,intercept=0,colour="red") + stat_smooth(method=lm) + theme_bw()
```

The above plot compares the predicted additive and clonal values. The slope of the regression line (blue) is less than 1 (shown in red) and represents the additive proportion of the clonal value.

Previous studies have shown that the reliability of genomic prediction depends largely on the number of close relationships for each clone, which can be quantified by the 95th (or other suitably high) percentile of its G coefficients. 

```{r, fig.height=3,fig.width=4}
tab <- table(pheno$id)
Gtmp <- G
diag(Gtmp) <- NA
data2 <- data.frame(id=rownames(G),
                    G95=apply(Gtmp,1,quantile,probs=0.95,na.rm=T),
                    plots=as.integer(tab[rownames(G)]))

combo <- merge(data2,add.pred)
ggplot(combo,aes(x=G95,y=r2,colour=plots)) + geom_point() +
  scale_y_continuous("Reliability",limits=c(0.4,1),breaks=seq(0.4,1,by=0.2),labels=seq(0.4,1,by=0.2))
```

Because these are *marker-assisted* predictions, meaning they also utilize phenotypic data for the selection candidates, the number of phenotype measurements (i.e., plots) for each clone also contributes to accuracy (see above figure). The average reliability of the predicted additive values for FY3 entries, based on a single plot, is 0.62 vs. only 0.31 for the narrow-sense $h^2$.

```{r}
#Average reliability based on one plot
mean(combo$r2[combo$plots==1])
```

polyBreedR has two ways to make *marker-based* predictions, i.e., for individuals without phenotype data. In a real situation (as opposed to cross-validation), the G matrix will contain individuals for which BLUEs are not available in the data frame provided to `Stage2`, and predictions for these individuals will be present in the output from `predict_MME`. For cross-validation exercises, it is more efficient to mask the phenotypes of individuals using the argument "mask" in `predict_MME`, rather than re-estimating variances for each partition of the dataset using `Stage2`. 

To illustrate, we will mask phenotypes for the clones in the 2020 preliminary yield trial (Field Year 3), which all start with the prefix "W17". 

```{r,fig.height=3,fig.width=4}
W17 <- blue$id[grep("W17",blue$id,fixed=T)]
add.pred.W17 <- predict_MME(data=ans2$MME,weights=c(I=0,G=1),
                            mask=data.frame(id=W17))

combo <- merge(add.pred.W17,add.pred,by="id")[,c("id","r2.x","r2.y")]
combo$group <- ifelse(combo$id %in% W17,"W17","Remainder") 
ggplot(combo,aes(x=r2.x,y=r2.y,colour=group)) + geom_point() + geom_abline(intercept=0,slope=1) + 
  xlab(expression(r^2~"(-W17 phenotypes)")) + ylab(expression(r^2~"(+W17 phenotypes)")) + scale_colour_brewer(palette="Set1")
```

The above figure compares the reliability of the predicted additive values with (y-axis) and without (x-axis) phenotypes for the W17 group (black line = slope of 1). The reliability is substantially higher for the W17 clones when their phenotype data are included, which reinforces the difference between marker-assisted and marker-based predictions. For the rest of the population, there is a slight advantage to including the W17 phenotypes because they increase the training population size.

To estimate accuracy by cross-validation, the correlation between the predicted and observed values for the W17 group is divided by the square-root of the narrow-sense $h^2$. The result of 0.49 is lower than the square-root of the mean reliability = 0.72, and the true value may be somewhere in between.

```{r}
#Predictive ability
pred.ability <- cor(blue$total.yield[match(W17,blue$id)],add.pred.W17$blup[match(W17,add.pred.W17$id)])

#Prediction accuracy by cross-validation
pred.ability/sqrt(0.31)

#Compared to square root of mean reliability
sqrt(mean(add.pred.W17$r2[match(W17,add.pred.W17$id)]))
```

